packages/contracts/contracts/TransactionManager.sol

```diff
@@ -416,34 +416,42 @@ contract TransactionManager is ReentrancyGuard, ProposedOwnable, ITransactionMan
+    { // scope: validation and effects
+      bytes32 digest = hashInvariantTransactionData(txData);
 
-    // Make sure the expiry has not elapsed
-    require(txData.expiry >= block.timestamp, "#F:020");
+      // Make sure that the variant data matches what was stored
+      require(variantTransactionData[digest] == hashVariantTransactionData(txData.amount, txData.expiry, txData.preparedBlockNumber), "#F:019");
 
-    // Make sure the transaction wasn't already completed
-    require(txData.preparedBlockNumber > 0, "#F:021");
+      // Make sure the expiry has not elapsed
+      require(txData.expiry >= block.timestamp, "#F:020");
 
-    // Validate the user has signed
-    require(recoverSignature(txData.transactionId, relayerFee, "fulfill", signature) == txData.user, "#F:022");
+      // Make sure the transaction wasn't already completed
+      require(txData.preparedBlockNumber > 0, "#F:021");
 
-    // Sanity check: fee <= amount. Allow `=` in case of only wanting to execute
-    // 0-value crosschain tx, so only providing the fee amount
-    require(relayerFee <= txData.amount, "#F:023");
+      // Validate the user has signed
+      require(recoverSignature(txData.transactionId, relayerFee, "fulfill", signature) == txData.user, "#F:022");
 
-    // Check provided callData matches stored hash
-    require(keccak256(callData) == txData.callDataHash, "#F:024");
+      // Sanity check: fee <= amount. Allow `=` in case of only wanting to execute
+      // 0-value crosschain tx, so only providing the fee amount
+      require(relayerFee <= txData.amount, "#F:023");
 
-    // To prevent `fulfill` / `cancel` from being called multiple times, the
-    // preparedBlockNumber is set to 0 before being hashed. The value of the
-    // mapping is explicitly *not* zeroed out so users who come online without
-    // a store can tell the difference between a transaction that has not been
-    // prepared, and a transaction that was already completed on the receiver
-    // chain.
-    variantTransactionData[digest] = hashVariantTransactionData(txData.amount, txData.expiry, 0);
+      // Check provided callData matches stored hash
+      require(keccak256(callData) == txData.callDataHash, "#F:024");
+
+      // To prevent `fulfill` / `cancel` from being called multiple times, the
+      // preparedBlockNumber is set to 0 before being hashed. The value of the
+      // mapping is explicitly *not* zeroed out so users who come online without
+      // a store can tell the difference between a transaction that has not been
+      // prepared, and a transaction that was already completed on the receiver
+      // chain.
+      variantTransactionData[digest] = hashVariantTransactionData(txData.amount, txData.expiry, 0);
+    }
+
+    // Declare these variables for the event emission. Are only assgined
+    // IFF there is an external call on the receiving chain
```

Comment:
nit : punctuation `.` sry

Reply:
I guess this doesn't matter, comments here are pretty grammatically inconsistent lol